File: src/lista.py
class No:
    def __init__(self, dado):
        self.dado = dado
        self.proximo = None

class ListaEncadeada:
    def __init__(self):
        self.cabeca = None

    def adicionar(self, dado, posicao):
        # Adiciona um nó com dado na posição especificada
        novo_no = No(dado)
        if posicao == 0:
            novo_no.proximo = self.cabeca
            self.cabeca = novo_no
            return

        # Percorrendo a lista para encontrar a posição válida
        atual = self.cabeca
        contador = 0

        while atual and contador < posicao - 1:
            atual = atual.proximo
            contador += 1

        # Verifica se a posição é válida (se 'atual' existe e está no índice correto)
        if atual is None:
            return

        # Inserção do novo nó na posição especificada
        novo_no.proximo = atual.proximo
        atual.proximo = novo_no

    def remover(self, dado):
    # Remove o primeiro nó encontrado com o valor dado
        atual = self.cabeca
        anterior = None

        while atual:
            if atual.dado == dado:
                if anterior:
                    anterior.proximo = atual.proximo
                else:
                    self.cabeca = atual.proximo

                # Remove referências ao nó
                atual.proximo = None  # Remove referência ao próximo nó
                atual.dado = None  # Remove o dado do nó
                return  # Sai após remover o nó encontrado
            anterior = atual
            atual = atual.proximo


    def exibir(self):
        #Exibe a lista encadeada
        elementos = []
        atual = self.cabeca
        while atual:
            elementos.append(atual.dado)
            atual = atual.proximo
        print("Lista: ", elementos)

def main():
    # Lendo o arquivo
    with open("src/arq-novo.txt") as arquivo:
        # Leitura da lista inicial
        valores_iniciais = list(map(int, arquivo.readline().split()))
        
        # Criando a lista encadeada e adicionando os elementos iniciais
        lista_encadeada = ListaEncadeada()
        for i, valor in enumerate(valores_iniciais):
            lista_encadeada.adicionar(valor, i)

        # Leitura do número de ações
        num_acoes = int(arquivo.readline().strip())

        # Executando as ações
        for _ in range(num_acoes):
            linha_acao = arquivo.readline().strip().split()
            tipo_acao = linha_acao[0]
            
            if tipo_acao == 'A':
                numero = int(linha_acao[1])
                posicao = int(linha_acao[2])
                lista_encadeada.adicionar(numero, posicao)
            elif tipo_acao == 'R':
                numero = int(linha_acao[1])
                lista_encadeada.remover(numero)
            elif tipo_acao == 'P':
                lista_encadeada.exibir()
main()

File: bin/lista.js

const fs = require('fs');
const readline = require('readline');

// Definindo a Classe para os Nós
class No {
    constructor(valor) {
        // variaveis com _ para não conflitar.
        this.valor_ = valor;
        this.proximo_ = null;
    }

    // Definindo os gets e sets

    get valor() {
        return this.valor_;
    }

    set valor(novoValor) {
        this.valor_ = novoValor;
    }

    get proximo() {
        return this.proximo_;
    }

    set proximo(novoProximo) {
        this.proximo_ = novoProximo;
    }
}

// Definindo a Classe para a Lista Ligada

class Lista {
    constructor() {
        this.cabeca = null;
    }

    // Metodo para inserir elementos no fim da lista

    inserirFim(valor) {
        let novoNo = new No(valor);
        if (this.cabeca === null) {
            this.cabeca = novoNo;
        } else {
            let atual = this.cabeca;
            while (atual.proximo !== null) {
                atual = atual.proximo;
            }
            atual.proximo = novoNo;
        }
    }

    // Metodo para inserir elementos no indice especificado

    inserirIndice(valor, indice) {
        let novoNo = new No(valor);

        if (indice === 0) {
            novoNo.proximo = this.cabeca;
            this.cabeca = novoNo;
            return;
        }

        let atual = this.cabeca;
        let contador = 0;

        while (atual !== null) {
            if (contador === indice - 1) {
                novoNo.proximo = atual.proximo;
                atual.proximo = novoNo;
                return;
            }
            atual = atual.proximo;
            contador++;
        }

        console.log("Índice fora dos limites.");
    }

    // Metodo para remover valores especificos

    remover(valor) {
        if (this.cabeca === null) {
            return;
        }

        if (this.cabeca.valor === valor) {
            this.cabeca = this.cabeca.proximo;
            return;
        }

        let anterior = this.cabeca;
        let atual = this.cabeca.proximo;

        while (atual !== null) {
            if (atual.valor === valor) {
                anterior.proximo = atual.proximo;
                return;
            }
            anterior = atual;
            atual = atual.proximo;
        }
    }

    // Metodo para imprimir toda a lista
    imprimir() {
        let atual = this.cabeca;
        let valores = [];
        while (atual !== null) {
            valores.push(atual.valor); // Adiciona cada valor ao array
            atual = atual.proximo;
        }
        console.log(valores.join(", ")); // Exibe os valores na horizontal
    }


}
// função para ler os arquivos txt e modificar a lista de acordo com as operações.
function lerArquivo(path, lista) {
    
    // criando a interface para leitura
    const rl = readline.createInterface({
        input: fs.createReadStream(path),
        crlfDelay: Infinity
    });

    let primeiraLinha = true;
    
    //confirmação para a leitura do arquivo
    console.log('Arquivo sendo lido');

    //leitura do arquivo
    rl.on('line', (linha) => {
        // para a primeira linha como não tem uma ação especificada, o codigo salva em um array
        // todos os numeros e adiciona 1 por 1 na lista.
        console.log('Nova linha lida:');
        if (primeiraLinha) {
            console.log('Primeira linha do arquivo');

            var numeros = [];

            numeros = linha.trim().split(' ').map(Number);
            console.log('Números sendo adicionados:', numeros);

            // Adiciona os números à lista
            for (let numero of numeros) {
                lista.inserirFim(numero);
            }

            // primeira linha false, o codigo verifica se tem operações demarcadas nas linhas
            primeiraLinha = false;
        } 
        else {

            // salvando as operações linha por linha em arrays e separando em array
            // operacao é toda a linha separada em array
            // comando é a String de operação. ex: A, P e R
            // em args é guardado o valor do No e se tiver na linha o Indice.

            console.log('Operação:', linha);
            const operacao = linha.trim().split(' ');
            const comando = operacao[0];
            const args = operacao.slice(1);

            // Executa as operações especificas na lista
            if (comando === 'P') {
                lista.imprimir();
            } else if (comando === 'A') {

                // args[0] vai puxar o valor do No no array args
                // args[1] vai puxar o indice no No no array args

                const valor = parseInt(args[0]);
                const indice = parseInt(args[1]);
                lista.inserirIndice(valor, indice);
            } else if (comando === 'R') {
                const valor = parseInt(args[0]);
                lista.remover(valor);
            }
        }
    });

    // fechando a interface de leitura do arquivo

    rl.on('close', () => {
        console.log('Fim da leitura do arquivo');
    });
}
    
const lista = new Lista();

lerArquivo("src\\arq-novo.txt", lista);

lista.imprimir();
